### PR TITLE
channel: introduce Resolve and ResolveLocked

### PR DESCRIPTION
**Pull request by pedronis**
This introduces helpers to combine a given channel/locked track with a new channel, preserving the track respectively making sure the track is kept if the new channel is risk/branch-only. Notice that ATM the new channel is not strictly parsed, it might be in the future where we have moved to sanitize our channel inputs more.

These are then used in image and snapstate to simplify/homogenize some code.

Channel.Verbatim(Track|Risk)Only -> bool are added too.

Some drive-bys as well.

### CODE REVIEW

File: image/image.go [hunk 1]
```diff
@@ -275,6 +275,11 @@ func snapChannel(name string, model *asserts.Model, opts *Options, local *localI
 		// fallback to default channel
 		snapChannel = opts.Channel
 	}
+	if snapChannel != "" {
+		if _, err := channel.ParseVerbatim(snapChannel, "_"); err != nil {
```
Comment (bboozzoo) [inline image/image.go]: We sometimes use `"-"` or `"_"` for arch to avoid the default one. Maybe as a drive-by, it'd be useful to add a const in the `snap/channel` package? eg.
```
const NoDefaultArchitecture = "-"
```

File: snap/channel/channel.go [hunk 2]
```diff
@@ -200,3 +211,53 @@ func (c *Channel) Match(c1 *Channel) ChannelMatch {
...
+	}
+	p := strings.Split(newChannel, "/")
+	// risk/branch only
+	if strutil.ListContains(channelRisks, p[0]) && ch.Track != "" {
+		return ch.Track + "/" + newChannel, nil
+	}
+	return newChannel, nil
+}
+
+var ErrTrackSwitch = errors.New("cannot switch locked track")
```
Comment (zyga) [inline snap/channel/channel.go]: Nitpick: the error name is a bit generic (any kind of track switch operation can claim this name). I'd suggest using something matching `Locked` name in `ResolveLocked`.

File: snap/channel/channel.go [hunk 3]
```diff
@@ -150,6 +151,16 @@ func (c *Channel) Full() string {
 	return c.String()
 }
 
+// VerbatimTrackOnly returns whether the channel represents a track only.
+func (c *Channel) VerbatimTrackOnly() bool {
```
Comment (chipaca) [inline snap/channel/channel.go]: I feel uncomfortable about the `Verbatim` in these. It feels like there's a separate type inside `Channel` trying to get out...

File: snap/channel/channel.go [hunk 4]
```diff
@@ -200,3 +211,53 @@ func (c *Channel) Match(c1 *Channel) ChannelMatch {
...
+	}
+	return newChannel, nil
+}
+
+var ErrTrackSwitch = errors.New("cannot switch locked track")
+
+// ResolveLocked resolves newChannel wrt a locked track, newChannel
+// can only be risk/branch-only or have the same track, otherwise
+// ErrTrackSwitch is returned.
+func ResolveLocked(track, newChannel string) (string, error) {
```
[the remark this answers is not in the document]
Comment (chipaca) [reply snap/channel/channel.go]: I understand this naming is part of an ongoing conversation, that we'll wrap up next week. As such, much as `Locked` irks me, I'm ok for it to go in to help settle this.

File: snap/channel/channel.go (hunk 3 above)
Comment (pedronis) [reply snap/channel/channel.go]: I agree, we would need to have different types for ParseVerbatim result vs Parse/Clean, *but* OTOH TBH it feels a bit overkill at this juncture.
